ostd/src/arch/x86/kernel/acpi/multiprocessor_wakeup.rs

```diff
@@ -0,0 +1,593 @@
+                    });
+                }
+
+                MadtEntry::LocalApicNmi(entry) => local_apic_nmi_lines.push(NmiLine {
+                    processor: if entry.processor_id == 0xff {
+                        NmiProcessor::All
+                    } else {
+                        NmiProcessor::ProcessorUid(entry.processor_id as u32)
+                    },
+                    line: match entry.nmi_line {
+                        0 => LocalInterruptLine::Lint0,
+                        1 => LocalInterruptLine::Lint1,
+                        _ => return Err(AcpiError::InvalidMadt(MadtError::InvalidLocalNmiLine)),
+                    },
+                }),
+
+                MadtEntry::X2ApicNmi(entry) => local_apic_nmi_lines.push(NmiLine {
+                    processor: if entry.processor_uid == 0xffffffff {
+                        NmiProcessor::All
+                    } else {
+                        NmiProcessor::ProcessorUid(entry.processor_uid)
+                    },
+                    line: match entry.nmi_line {
+                        0 => LocalInterruptLine::Lint0,
+                        1 => LocalInterruptLine::Lint1,
+                        _ => return Err(AcpiError::InvalidMadt(MadtError::InvalidLocalNmiLine)),
+                    },
+                }),
+
+                MadtEntry::LocalApicAddressOverride(entry) => {
+                    local_apic_address = entry.local_apic_address;
+                }
+
+                MadtEntry::MultiprocessorWakeup(_) => {}
+
+                _ => {
+                    return Err(AcpiError::InvalidMadt(MadtError::UnexpectedEntry));
+                }
+            }
+        }
+
+        Ok((
+            InterruptModel::Apic(Apic {
+                local_apic_address,
+                io_apics,
+                local_apic_nmi_lines,
+                interrupt_source_overrides,
+                nmi_sources,
+                also_has_legacy_pics: self.supports_8259(),
+            }),
+            Some(ProcessorInfo {
+                boot_processor: boot_processor.unwrap(),
+                application_processors,
+            }),
+        ))
+    }
+
+    pub fn entries(&self) -> MadtEntryIter {
+        MadtEntryIter {
+            pointer: unsafe { (self as *const Madt as *const u8).add(mem::size_of::<Madt>()) },
+            remaining_length: self.header.length - mem::size_of::<Madt>() as u32,
+            _phantom: PhantomData,
+        }
+    }
+
+    pub fn supports_8259(&self) -> bool {
+        { self.flags }.get_bit(0)
+    }
+}
+
+pub struct MadtEntryIter<'a> {
+    pointer: *const u8,
+    /*
+     * The iterator can only have at most `u32::MAX` remaining bytes, because the length of the
+     * whole SDT can only be at most `u32::MAX`.
+     */
+    remaining_length: u32,
+    _phantom: PhantomData<&'a ()>,
+}
+
+#[allow(dead_code)]
+pub enum MadtEntry<'a> {
+    LocalApic(&'a LocalApicEntry),
+    IoApic(&'a IoApicEntry),
```

Comment:
Are the type definitions copied from the `apic` crate? If so, we need to add some license header. Take a look at [how we have added license headers for the code copied from `trapframe-rs` crate](https://github.com/asterinas/asterinas/blob/897661e30e03bed17acfcf111040d4c33bb8d09f/ostd/src/arch/x86/trap/syscall.S#L1-L16) as an example.